README.md
# Stella - 2.1-RC1

**Define realistic testplans and run them against your webapps**

## Features

## Examples

### Testplan Configuration

### Running Tests

## Installation

Get it in one of the following ways:
     
    $ gem install stella
    $ sudo gem install stella
    $ git clone git://github.com/solutious/stella.git

You can also download via [tarball](http://github.com/solutious/stella/tarball/latest) or [zip](http://github.com/solutious/stella/zipball/latest). 

NOTE: If you get errors about libxml2 or libxslt on Ubuntu, you need to install the following:

    sudo apt-get install libxml2-dev
    sudo apt-get install libxslt1-dev


## More Information

* [Homepage](http://solutious.com/projects/stella)
* [Codes](http://github.com/solutious/stella)
* [RDocs](http://solutious.com/stella)
* [Stellaaahhhh](http://stellaaahhhh.com)
* [Using Stella for Web application testing](http://searchsoftwarequality.techtarget.com/tip/0,289483,sid92_gci1510488,00.html)


## Credits

* [Delano Mandelbaum](http://solutious.com)
* [Daniel J. Berger] for file logic borrowed from ptools[http://raa.ruby-lang.org/project/ptools/]


## Thanks 

* Harm Aarts for the great test case and feedback!
* Kalin Harvey for keeping me on track.
* Dave L, the best intern money can't buy. 
* Peter McCurdy for the feedback and bug fixes. 


## Related Projects

* [WatirGrid](http://github.com/90kts/watirgrid)
* [Watir](http://watir.com/)
* [JMeter](http://jakarta.apache.org/jmeter/)
* [Tsung](http://tsung.erlang-projects.org/)
* [Grinder](http://grinder.sourceforge.net/)
* [Pylot](http://www.pylot.org/)
* [Trample](http://github.com/jamesgolick/trample)

## License

See LICENSE.txt


CHANGES.txt
STELLA, CHANGES

#### 2.1.2 (2011-07-06) ###############################

* FIXED: Stella::Report#postprocess loads log from a hash.
* ADDED: Wait only up to 35 seconds for a remote checkup
* CHANGE: Move stella/api.rb content to stella.rb
* CHANGE: CLI output
* CHANGE: Removed Stella::Template and require 'erb'


#### 2.1.1 (2011-07-06) ###############################

* ADDED: Append more info URI to CLI output for remote checkups

#### 2.1.0 (2011-07-06) ###############################

* FIXED: Different CLI output when a testplan is specified
* ADDED: Handle bad or missing api credentials
* ADDED: Globals -A and -K for supplying API credentials
* CHANGE: Default output is now CLI (use -F json)
* CHANGE: Global -f is now -u (for usecase)
* CHANGE: Global -F is now -f (output format)
* CHANGE: STELLA_USER env var is now STELLA_ACCOUNT


#### 2.0.3 (2011-07-06) ###############################

* ADDED: Stella::API
* ADDED: --remote global for running checkups via www.blamestella.com (beta)


#### 2.0.2 (2011-06-09) ###############################

* FIXED: Gibbler ID fixes
* ADDED: --wait option for checkup command
* CHANGE: Don't freeze testplan


#### 2.0.1 (2011-06-08) ###############################

NOTE: First release of 2.x.


#### 0.8.X (XXXX-XX-XX) ###############################

NOTE: We're now focusing our effort on 2.0, so these features are still 
      outstanding for a 1.x release (they probably won't be worked on).

* TODO: Force response block content type. 
* TODO: Add option to dump full request/response output to a file by client
* TODO: review cookie handling. Not always sent automatically. 
* TODO: review :variables in URI elements
* TODO: global responses
* TODO: request block conditions. 
* TODO: process templates for calls to set in get blocks

#### 0.8.8 (2011-02-26) ###############################

* FIXED: Require benelux 0.5.17 (fixed for Ruby 1.9.2)

#### 0.8.7.002 (2010-06-14) ###############################

* FIXED: Specifying a host w/ a path and no testplan no longer doubles up the path
* ADDED: Global HTTP client timeout ("stella -T 60")
* CHANGE: Timeouts are reported as exceptions. 


#### 0.8.7 (2010-05-21) ###############################

* FIXED: sequential, rsequential, and random values were broken for Arrays
* FIXED: Support for supplying test durations in the form "5m" (300 seconds)
* CHANGE: Enforce port 443 if port 80 and scheme is HTTPS. 


#### 0.8.6 (2010-04-14) ###############################

* FIXED: Wait option and hosts on CLI were not being passed correctly
* CHANGE: Stella::Testplan#userid now used in digest calculation.
* CHANGE: Moved ProcSource to Storable
* CHANGE: Don't auto freeze testrun.
* CHANGE: Store Stella::Data::HTTP::Request#wait as a Range
* ADDED: Stella::Testplan#userid
* ADDED: Stella.post
* ADDED: Stella::Logger.debug
* ADDED: Print wait time for each request at verbose level 3 (-vvv)


#### 0.8.5 (2010-03-27) ###############################

* FIXED: Add User-Agent header if Stella.agent is not null and one is not already provided
* FIXED: Client wait (undefined method '>' for Range error)
* CHANGE: Testrun#samples contains all data samples for each event in a given batch
* CHANGE: Renamed Testrun#events -> Testrun#event_probes
* ADDED: Testrun#logsize
* ADDED: Store HTTP status codes in Testrun#stats
* ADDED: Store error messages in Sample#errors
* ADDED: Stella.get now takes a block
* ADDED: Stella.agent


#### 0.8.4 (2010-03-19) ###############################

* FIXED: Misc issues getting and setting headers, params, and resources
* FIXED: There was no response body in Stella::Engine::Log objects
* CHANGE: Store URIs as strings in Stella::Engine::Log objects
* CHANGE: Stella::Data::HTTP now uses @header instead of @headers
* ADDED: Testrun#etime, Testrun#running?
* ADDED: Testrun.elapsed_seconds
* ADDED: Testrun.granularity which replaces Load::ROTATE_TIMELINE


#### 0.8.3 (2010-03-15) ###############################

* FIXED: Testrun digest handling
* FIXED: Testrun.id will be converted to a Gibbler::Digest if it's a String. 
* FIXED: Reconstituting Testrun from a hash (samples and stats)
* CHANGE: Removed dashes from CLI params: notemplate, nostats, nowait, withheader, withparam
* CHANGE: Engine is now called via Testrun#run
* CHANGE: Use the ID attribute (e.g. Testplan.id) instead of digest or digest_cache (which only
  be used once to generate an ID 
* ADDED: Testrun#client_options
* ADDED: Reconstituting Stella::Testrun and Stella::Testplan objects from JSON/Hash
* ADDED: Testrun objects are now frozen after the run
* ADDED: Testrun#userid, Testrun#runinfo
* ADDED: Default Testrun status is "new"
* ADDED: Testrun#run now accepts an options hash which is passed to Engine#run
* ADDED: Engine option: with_content (populate Testrun#log during a load test)
* ADDED: First batch of requests store header content (for load tests)


#### 0.8.2 (2010-03-05) ###############################

* FIXED: Misc issues with usage via module (Stella::Engine.load)
* FIXED: ANSI code method stubs only created when Drydock not available
* CHANGE: Remove all logs except summary and exceptions
* CHANGE: Renamed .stella/logs/latest/request-exceptions to .stella/logs/latest/exceptions
* CHANGE: Removed unused Load implementations and renamed LoadQueue to Load
* CHANGE: Added proper names for statistics (response_time instead of do_request)
* CHANGE: Benelux 0.5.8 dependency
* CHANGE: Cleaned up generate mode CLI output
* CHANGE: Removed no-logging option
* CHANGE: Renamed Testrun#summary to Testrun#stats
* CHANGE: Pass entire URI when calling Testplan#new
* CHANGE: Engine objects are now classes
* CHANGE: Engine#run methods now return instance of Stella::Testrun
* ADDED: JSON statistics log (.stella/logs/latest/stats)


#### 0.8.1.002 (2010-02-20) ###############################

* FIXED: fail on connection error
* CHANGE: Remove hanna dependency
* CHANGE: Now store complete headers sent rather than just configured. 



lib/stella.rb
# encoding: utf-8
STELLA_LIB_HOME = File.expand_path File.dirname(__FILE__) unless defined?(STELLA_LIB_HOME)

if RUBY_VERSION =~ /1.9/
  Encoding.default_external = Encoding::UTF_8
  Encoding.default_internal = Encoding::UTF_8
end

require 'bundler/setup'
$:.unshift STELLA_LIB_HOME

%w{tryouts benelux storable gibbler familia bluth}.each do |dir|
  $:.unshift File.join(STELLA_LIB_HOME, '..', '..', dir, 'lib')
end

require 'yajl'
require 'httparty'
require 'storable'
require 'benelux'
require 'gibbler/aliases'
require 'stella/core_ext'
require 'familia'

autoload :Nokogiri, 'nokogiri'

class Stella
  module VERSION
    def self.to_s
      load_config
      [@version[:MAJOR], @version[:MINOR], @version[:PATCH]].join('.')
    end
    def self.inspect
      load_config
      [@version[:MAJOR], @version[:MINOR], @version[:PATCH], @version[:BUILD]].join('.')
    end
    def self.load_config
      require 'yaml'
      @version ||= YAML.load_file(File.join(STELLA_LIB_HOME, '..', 'VERSION.yml'))
    end
  end
end

#
# Any object that wants to be serialized to JSON 
# ought to inherit from this class. 
# 
# NOTE: you cannot define Storable fields here.
# Most notably, you'll prob want to include this. 
#
#   field :id, :class => Gibbler::Digest, :meth => :gibbler, &gibbler_id_processor
#
#module StellaObject < Storable
#  include Gibbler::Complex
#  def id
#    @id ||= self.digest
#    @id
#  end
#end

# All errors inherit from this class. 
class StellaError < RuntimeError
  def initialize(msg)
    @message = msg
  end
  attr_reader :message
end

class StellaBehavior < Exception
  def initialize(msg)
    @message = msg
  end
  attr_reader :message
end

class Stella
  class ForcedRedirect < StellaBehavior
    attr_accessor :location
    def initialize(l)
      @location = l
    end
  end
  class RepeatRequest < StellaBehavior
  end
  class UsecaseFail < StellaBehavior
  end
  class TestplanQuit < StellaBehavior
  end
  class RequestError < StellaBehavior
  end
  class TimeoutError < StellaError
  end
  class UsecaseError < StellaError
  end
  class PageError < StellaError
  end
  class HTTPError < StellaError
    attr_reader :status
    def initialize(status)
      @status = status
    end
    def message
      "#{status} error"
    end
  end
  require 'stella/utils'
  require 'stella/client'
  require 'stella/engine'
  require 'stella/report'
  require 'stella/testplan'
  attr_reader :plan
  def initialize *args
    @plan = Stella::TP === args.first ? 
      args.first.clone : Stella::TP.new(args.first)
    #@plan.freeze
    @runner
  end
end 


class Stella
  @sysinfo = nil
  @debug   = false
  @noise   = 1
  @abort   = false
  @quiet   = false
  @agent   = "Mozilla/5.0 (compatible; Stella/#{Stella::VERSION}; +http://solutious.com/projects/stella)"  
  # static methods
  class << self
    attr_accessor :log, :stdout, :agent, :debug, :quiet, :noise
    def debug?()        @debug == true  end
    def quiet?()        @quiet == true  end
    def li(*msg) STDERR.puts *msg unless quiet? end
    def le(*msg); li "  " << msg.join("#{$/}  ") end
    def ld(*msg)
      return unless Stella.debug?
      prefix = "D(#{Thread.current.object_id}):  "
      li("#{prefix}#{msg.join("#{$/}#{prefix}")}")
    end
    
    def get(uri, opts={})
      opts[:concurrency] ||= 1
      opts[:repetitions] ||= 1
      run = checkup uri, opts
      report = run.report
      if report.processed? && report.content && report.statuses.success?
        report.content.response_body 
      else
        nil
      end
    end
    def checkup(uri, opts={})
      plan = Stella::Testplan.new uri
      run = Stella::Testrun.new plan, :checkup, opts
      Stella::Engine.run run
      run
    end
    def now
      Time.now.utc.to_f
    end
    # http://blamestella.com/ => blamestella.com
    # https://blamestella.com/ => blamestella.com:443
    def canonical_host(host)
      return nil if host.nil?
      if host.kind_of?(URI)
        uri = host
      else
        host &&= host.to_s
        host.strip!
        host = host.to_s unless String === host
        host = "http://#{host}" unless host.match(/^https?:\/\//)
        uri = URI.parse(host)
      end
      str = "#{uri.host}".downcase 
      #str << ":#{uri.port}" if uri.port && uri.port != 80 
      str
    end
  
    def canonical_uri(uri)
      return nil if uri.nil?
      if uri.kind_of?(URI)
        uri = Addressable::URI.parse uri.to_s
      elsif uri.kind_of?(String)
        uri &&= uri.to_s
        uri.strip! unless uri.frozen?
        uri = "http://#{uri}" unless uri.match(/^https?:\/\//)
        uri = Addressable::URI.parse(uri)
      end
      uri.scheme ||= 'http'
      uri.path = '/' if uri.path.to_s.empty?
      uri
    end
    
    def rescue(&blk)
      blk.call
    rescue StellaError => ex
      Stella.le ex.message
      Stella.ld ex.backtrace
    #rescue => ex
    #  Stella.le ex.message
    #  Stella.le ex.backtrace
    end
    
  end
end

class Stella  
  class API
    include HTTParty
    ssl_ca_file Stella::Client::SSL_CERT_PATH
    format :json
    attr_reader :httparty_opts, :response, :account, :key
    def initialize account=nil, key=nil, httparty_opts={}
      self.class.base_uri ENV['STELLA_HOST'] || 'https://www.blamestella.com/api/v2'
      @httparty_opts = httparty_opts
      @account = account || ENV['STELLA_ACCOUNT']
      @key = key || ENV['STELLA_KEY']
      unless @account.to_s.empty? || @key.to_s.empty?
        httparty_opts[:basic_auth] ||= { :username => @account, :password => @key }
      end
    end
    def get path, params=nil
      opts = httparty_opts
      opts[:query] = params || {}
      execute_request :get, path, opts
    end
    def post path, params=nil
      opts = httparty_opts
      opts[:body] = params || {}
      execute_request :post, path, opts
    end
    def site_uri path
      uri = Addressable::URI.parse self.class.base_uri
      uri.path = uri_path(path)
      uri.to_s
    end
    private
    def uri_path *args
      args.unshift ''  # force leading slash
      path = args.flatten.join('/')
      path.gsub '//', '/'
    end
    def execute_request meth, path, opts
      path = uri_path [path]
      @response = self.class.send meth, path, opts
      indifferent_params @response.parsed_response
    end
    # Enable string or symbol key access to the nested params hash.
    def indifferent_params(params)
      if params.is_a?(Hash)
        params = indifferent_hash.merge(params)
        params.each do |key, value|
          next unless value.is_a?(Hash) || value.is_a?(Array)
          params[key] = indifferent_params(value)
        end
      elsif params.is_a?(Array)
        params.collect! do |value|
          if value.is_a?(Hash) || value.is_a?(Array)
            indifferent_params(value)
          else
            value
          end
        end
      end
    end
    # Creates a Hash with indifferent access.
    def indifferent_hash
      Hash.new {|hash,key| hash[key.to_s] if Symbol === key }
    end
    
    class Unauthorized < RuntimeError
    end
  end
end


